Sim/SimDelphesInterface/src/DelphesSimulation.cpp

```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+              << " Rel E: "    << setw(10) << simE << " <-> " << recE;
+    std::cout << std::endl;*/
+  } // For - tracks
 }
 
-void DelphesSimulation::ConvertMET(   TObjArray *  Input , fcc::ParticleCollection *  coll  ){
+void DelphesSimulation::ConvertTowers(const TObjArray*  Input,
+                                      fcc::ParticleCollection*   colParticles,
+                                      fcc::MCParticleCollection* colMCParticles,
+                                      fcc::ParticleMCParticleAssociationCollection* ascColParticlesToMC)
+{
+  Candidate* cand = nullptr;
+  for(int j=0; j<Input->GetEntries(); j++) {
 
-  Candidate * cand;
-  for(int j = 0; j < Input->GetEntries(); ++j) {
+    cand = static_cast<Candidate *>(Input->At(j));
+
+    auto particle = colParticles->create();
+
+    auto barePart      = new fcc::BareParticle();
+    barePart->Type     = cand->PID;
+    barePart->Status   = cand->Status;
+    barePart->P4.Px    = cand->Momentum.Px();
+    barePart->P4.Py    = cand->Momentum.Py();
+    barePart->P4.Pz    = cand->Momentum.Pz();
+    barePart->P4.Mass  = cand->Momentum.M();
+    barePart->Charge   = cand->Charge;
+    barePart->Vertex.X = cand->Position.X();
+    barePart->Vertex.Y = cand->Position.Y();
+    barePart->Vertex.Z = cand->Position.Z();
+    particle.Core(*barePart);
+
+    // Test print
+    /*double energy = sqrt(particle.Core().P4.Px*particle.Core().P4.Px +
+                         particle.Core().P4.Py*particle.Core().P4.Py +
+                         particle.Core().P4.Pz*particle.Core().P4.Pz +
+                         particle.Core().P4.Mass*particle.Core().P4.Mass);
+    std::cout << "Delphes Tower: "
+              << " Id: "       << setw(3)  << j+1
+              << " Pdg: "      << setw(5)  << particle.Core().Type
+              << " Stat: "     << setw(2)  << particle.Core().Status
+              << " Bits: "     << setw(2)  << particle.Core().Bits
+              << " Px: "       << setw(10) << particle.Core().P4.Px
+              << " Py: "       << setw(10) << particle.Core().P4.Py
+              << " Pz: "       << setw(10) << particle.Core().P4.Pz
+              << " E: "        << setw(10) << energy
+              << " M: "        << setw(10) << particle.Core().P4.Mass
+              << " Vx: "       << setw(10) << particle.Core().Vertex.X
+              << " Vy: "       << setw(10) << particle.Core().Vertex.Y
+              << " Vz: "       << setw(10) << particle.Core().Vertex.Z
+              << std::endl;*/
+
+    // Reference to MC - Delphes holds references to all objects related to the Photon object, several relations might exist for gammas
+    std::set<int> idRefMCPart; // Avoid double counting when referencingh MC particles
+
+    Candidate* refCand  = nullptr;
+    Candidate* ref2Cand = nullptr;
+
+    TIter it1(cand->GetCandidates());
+    it1.Reset();
+    while((refCand = static_cast<Candidate*>(it1.Next()))) {
```

Comment:
see 
http://www.cprogramming.com/tutorial/stl/iterators.html
("The STL approach")
plus the second link above.